db/tables.py

```diff
@@ -358,3 +358,26 @@ def infer_table_column_types(
             column_name,
             engine,
         )
+
+
+def infer_table_column_types(schema, table_name, engine):
+    metadata = MetaData(bind=engine, schema=schema)
+    table = reflect_table(table_name, schema, engine)
+
+    temp_name = "temp_table"
+    temp_full_name = schema + "." + temp_name
+    columns = [c.copy() for c in table.columns]
+    temp_table = Table(temp_name, metadata, *columns)
+
+    create_table = f"""
+    CREATE TABLE {temp_full_name} AS
+    TABLE {table.schema}.{table.name}
+    """
+    with engine.begin() as conn:
+        conn.execute(DDL(create_table))
```

Comment:
For creating the temporary table, I suggest using the recipe posted by the SQLAlchemy maintainer here:  https://github.com/sqlalchemy/sqlalchemy/issues/5687 , but modified with the `TEMPORARY` prefix.  This will let you quickly create the table with the data you want to use for inference.

Your goal should be to end up with SQL along the lines of:

```sql
CREATE TEMPORARY TABLE my_temp_table AS SELECT * FROM my_orig_table;
```

The `TEMPORARY` prefix will result in the table being automatically dropped either at the end of the current transaction block, or whenever the connection is closed (when using the context manager, this should be when the outermost context-management block ends).  One caveat:  Temporary tables _cannot_ be created in a schema (they exist in a system schema).  For details about this schema caveat, and how to set the dropping behavior properly, see the docs here:  https://www.postgresql.org/docs/13/sql-createtable.html .

Reply:
An alternative route would be to try to copy the table in SQLAlchemy (i.e., reflect the table, and then copy the object).  You'd then do an `insert().from_select(...)` sort of statement:  https://docs.sqlalchemy.org/en/14/core/dml.html?highlight=from_select#sqlalchemy.sql.expression.Insert.from_select .  

I suspect that won't be as nice in the end, and it'll be a bit less efficient.  If you want to try, you can use `prefixes=["TEMPORARY"]` in your temp table definition, and then use the `table.create(engine)` method of the SA table object (see `create_mathesar_table` for an example of idiomatic table creation).  Once the table is created, use the `insert().from_select(...)` statement.

Reply:
I updated to use the `DDLElement` and `MathesarColumns`, but am struggling with getting the proper temporary table implementation working. Two issues so far:
* Problems reflecting the table after the type has been changed. Ideally we should be able to do this with `Table(autoload_with=conn)`, but I'm running into a `a transaction action has already begun` error. Making sure to run the type inference outside the initial `with conn.begin()` block, so not entirely sure whats happening here. As a temporary fix, we just pass the `MetaData` through all the inference functions, but this leads to cycles since the sqlalchemy table is never updated. 
* An error in the middle of the type inference functions seems to prevent the table from being dropped. When the above error occurs and our first test fails, the subsequent tests also fail with a `temp_table already exists` error. Will add a test for this behavior, but not entirely sure how to fix yet.

If there's anything obvious I'm doing wrong, please let me know! Otherwise, I'll keep hacking away at this tomorrow.